### PR TITLE
WIP : [tsachi] fast catchup

### PR DESCRIPTION
**Pull request by tsachiherman**
This PR is a WIP. It's still long way before it can be properly tested, and it's here just so I can ensure that newly added files aren't forgotten.

### CODE REVIEW

File: catchup/ledgerFetcher.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	return &ledgerFetcher{
+		net:      net,
+		accessor: accessor,
+		log:      log,
+		reporter: reporter,
+	}
+}
+
+func (lf *ledgerFetcher) getLedger(ctx context.Context, round basics.Round) error {
+	if len(lf.peers) == 0 {
```
Comment (pzbitskiy) [inline catchup/ledgerFetcher.go:60]: lock?

File: config/local_defaults.go [hunk 2]
```diff
@@ -39,6 +39,62 @@ const configVersion = uint32(6)
...
+	TelemetryToLog:                        true,
+	TxPoolExponentialIncreaseFactor:       2,
+	TxPoolSize:                            15000,
+	TxSyncIntervalSeconds:                 60,
+	TxSyncTimeoutSeconds:                  30,
+	TxSyncServeResponseSize:               1000000,
+	PeerConnectionsUpdateInterval:         3600,
+	DNSSecurityFlags:                      0x01,
+	EnablePingHandler:                     true,
+	CatchpointInterval:                    20000,
```
Comment (pzbitskiy) [inline config/local_defaults.go:94]: I like seeing comments near new/changed fields

File: util/db/dbutil.go [hunk 3]
```diff
@@ -83,28 +87,38 @@ func makeAccessorImpl(dbfilename string, readOnly bool, inMemory bool, params []
...
+	var dbHandler *sql.DB
+	dbHandler, sqliteConnections.initilizationError = sql.Open("sqlite3", URI("algorand-sqlite3-database-initializer", false, true)+"&_journal_mode=wal")
+	if sqliteConnections.initilizationError != nil {
+		return
+	}
+	sqliteConnections.initilizationError = dbHandler.Ping()
+	if sqliteConnections.initilizationError != nil {
+		return
+	}
+	sqliteConnections.initilizationError = dbHandler.Close()
```
Comment (pzbitskiy) [inline util/db/dbutil.go:101]: are you sure Close does not leak into sqlite3_uninitialize() somehow?

File: util/db/dbutil.go [hunk 4]
```diff
@@ -83,28 +87,38 @@ func makeAccessorImpl(dbfilename string, readOnly bool, inMemory bool, params []
...
-		if err != nil {
-			// init failed, db closed and err is set
-			return db, err
-		}
-		err = db.runInitStatements()
+
+	// we create a dummy in-memory database and close it right after. This would suffice to initialize the sqlite library threading logic.
+	var dbHandler *sql.DB
+	dbHandler, sqliteConnections.initilizationError = sql.Open("sqlite3", URI("algorand-sqlite3-database-initializer", false, true)+"&_journal_mode=wal")
+	if sqliteConnections.initilizationError != nil {
```
Comment (pzbitskiy) [inline util/db/dbutil.go:94]: initilizationError -> initializationError

File: agreement/gossip/network.go [hunk 5]
```diff
@@ -63,13 +63,16 @@ func WrapNetwork(net network.GossipNode, log logging.Logger) agreement.Network {
 	i.net = net
 	i.log = log
 
+	return i
+}
+
+func (i *networkImpl) Start() {
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply agreement/gossip/network.go:69]: ahh.. good question. We need this in order to be able to shut down the components without shutting down the network.

File: catchup/ledgerFetcher.go (hunk 1 above)
Comment (tsachiherman) [reply catchup/ledgerFetcher.go:60]: why ? do you see more than a single concurrent client ?

File: util/db/dbutil.go (hunk 3 above)
Comment (tsachiherman) [reply util/db/dbutil.go]: change was discarded.

File: util/db/dbutil.go (hunk 4 above)
Comment (tsachiherman) [reply util/db/dbutil.go]: change was discarded

**Pull request closed without merging**